cdap-ui/app/cdap/components/DataPrep/index.js

```diff
@@ -114,11 +96,33 @@ export default class DataPrep extends Component {
     });
     this.eventEmitter.off('DATAPREP_BACKEND_DOWN', this.toggleBackendDown);
     this.eventEmitter.off('DATAPREP_CLOSE_SIDEPANEL', this.closeSidePanel);
+    this.eventEmitter.off('REFRESH_DATAPREP', this.refreshDataPrep);
     if (this.dataprepStoreSubscription) {
       this.dataprepStoreSubscription();
     }
   }
 
+  refreshDataPrep = () => {
+    this.setState({
+      loading: true
+    });
+    /*
+      Not sure if this is necessary but added it is safer when doing an upgrade.
+      - Modified directives?
+      - Modified API calls that are not compatible with earlier version?
+    */
+    DataPrepStore.dispatch({
+      type: DataPrepActions.reset
+    });
+    let workspaceId = this.props.workspaceId;
+    this.setCurrentWorkspace(workspaceId);
+    setTimeout(() => {
```

Comment:
the `this.setCurrentWorkspace` already has setting the loading state to false. If there are cases where the loading didn't get set correctly, maybe add a `completed` callback function on the api call of setCurrentWorkspace